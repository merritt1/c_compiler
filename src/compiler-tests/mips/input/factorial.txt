void syscall_print_int(int integer);
void syscall_print_string(char *string);
int syscall_read_int(void);
void syscall_read_string(char *buffer, int length);
void syscall_exit(void);

int factorial(int n);

int main(void) {
  int n, f;

  syscall_print_string("Factorial of ");

  n = syscall_read_int();

  syscall_print_string(" equals ");

  f = factorial(n);

  syscall_print_int(f);

  syscall_print_string("\n");

  return 0;
}

int factorial(int n) {
  if(n <= 1)
    return 1;
  else
    return n*factorial(n-1);
}

