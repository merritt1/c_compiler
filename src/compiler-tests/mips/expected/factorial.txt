	.data
_StringLabel_2: .asciiz "\n"
_StringLabel_1: .asciiz " equals "
_StringLabel_0: .asciiz "Factorial of "

	.text
	.globl main
main:
	addiu $sp, $sp, -112
	sw $fp, 48($sp)
	or $fp, $sp, $0
	sw $ra, 52($fp)
	sw $a0, 0($fp)
	sw $a1, 4($fp)
	sw $a2, 8($fp)
	sw $a3, 12($fp)
	sw $s0, 16($fp)
	sw $s1, 20($fp)
	sw $s2, 24($fp)
	sw $s3, 28($fp)
	sw $s4, 32($fp)
	sw $s5, 36($fp)
	sw $s6, 40($fp)
	sw $s7, 44($fp)
	sw $a0, 96($fp)
	sw $a1, 100($fp)
	la $s0, _StringLabel_0
	or $a0, $s0, $0
	li $v0,4
	syscall
	la $s1, 96($fp)
	li $v0,5
	syscall
	or $s2, $v0, $0
	sw $s2, 0($s1)
	la $s3, _StringLabel_1
	or $a0, $s3, $0
	li $v0,4
	syscall
	la $s4, 100($fp)
	la $s5, 96($fp)
	lw $s6, 0($s5)
	or $a0, $s6, $0
	sw $t0, 56($fp)
	sw $t1, 60($fp)
	sw $t2, 64($fp)
	sw $t3, 68($fp)
	sw $t4, 72($fp)
	sw $t5, 76($fp)
	sw $t6, 80($fp)
	sw $t7, 84($fp)
	sw $t8, 88($fp)
	sw $t8, 92($fp)
	jal _Global_factorial
	lw $t9, 92($fp)
	lw $t8, 88($fp)
	lw $t7, 84($fp)
	lw $t6, 80($fp)
	lw $t5, 76($fp)
	lw $t4, 72($fp)
	lw $t3, 68($fp)
	lw $t2, 64($fp)
	lw $t1, 60($fp)
	lw $t0, 56($fp)
	or $s7, $v0, $0
	sw $s7, 0($s4)
	la $t0, 100($fp)
	lw $t1, 0($t0)
	or $a0, $t1, $0
	li $v0,1
	syscall
	la $t2, _StringLabel_2
	or $a0, $t2, $0
	li $v0,4
	syscall
	li $t3, 0
	or $v0, $t3, $0
	b _GeneratedLabel_1
_GeneratedLabel_1:
	li $v0,10
	syscall
_Global_factorial:
	addiu $sp, $sp, -104
	sw $fp, 48($sp)
	or $fp, $sp, $0
	sw $ra, 52($fp)
	sw $a0, 0($fp)
	sw $a1, 4($fp)
	sw $a2, 8($fp)
	sw $a3, 12($fp)
	sw $s0, 16($fp)
	sw $s1, 20($fp)
	sw $s2, 24($fp)
	sw $s3, 28($fp)
	sw $s4, 32($fp)
	sw $s5, 36($fp)
	sw $s6, 40($fp)
	sw $s7, 44($fp)
	sw $a0, 96($fp)
	la $s1, 96($fp)
	li $s2, 1
	lw $s3, 0($s1)
	sle $s4, $s3, $s2
	beqz $s4, _GeneratedLabel_2
	li $s5, 1
	or $v0, $s5, $0
	b _GeneratedLabel_3
_GeneratedLabel_2:
	bnez $s4, _GeneratedLabel_4
	la $s6, 96($fp)
	la $s7, 96($fp)
	li $t0, 1
	lw $t1, 0($s7)
	subu $t2, $t1, $t0
	or $a0, $t2, $0
	sw $t0, 56($fp)
	sw $t1, 60($fp)
	sw $t2, 64($fp)
	sw $t3, 68($fp)
	sw $t4, 72($fp)
	sw $t5, 76($fp)
	sw $t6, 80($fp)
	sw $t7, 84($fp)
	sw $t8, 88($fp)
	sw $t8, 92($fp)
	jal _Global_factorial
	lw $t9, 92($fp)
	lw $t8, 88($fp)
	lw $t7, 84($fp)
	lw $t6, 80($fp)
	lw $t5, 76($fp)
	lw $t4, 72($fp)
	lw $t3, 68($fp)
	lw $t2, 64($fp)
	lw $t1, 60($fp)
	lw $t0, 56($fp)
	lw $t3, 0($s6)
	or $t4, $v0, $0
	mul $t5, $t3, $t4
	or $v0, $t5, $0
	b _GeneratedLabel_3
_GeneratedLabel_4:
_GeneratedLabel_3:
	lw $s7, 44($fp)
	lw $s6, 40($fp)
	lw $s5, 36($fp)
	lw $s4, 32($fp)
	lw $s3, 28($fp)
	lw $s2, 24($fp)
	lw $s1, 20($fp)
	lw $s0, 16($fp)
	lw $ra, 52($fp)
	lw $fp, 48($fp)
	addiu $sp, $sp, 104
	jr $ra
